api/services/registry/server.go

```diff
@@ -0,0 +1,140 @@
+package earthly_registry_v1
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// NewServer creates and returns a new proxy server with a given host and client.
+func NewServer(addr string) *Server {
+	return &Server{
+		addr: addr,
+		cl:   &http.Client{},
+	}
+}
+
+type Server struct {
+	addr string
+	cl   *http.Client
+	UnimplementedRegistryServer
+}
+
+type streamSource interface {
+	Send(*ByteMessage) error
+	Recv() (*ByteMessage, error)
+}
+
+// NewStreamRW creates and returns a gRPC stream reader-writer that implements
+// io.Reader & io.Writer as to utilize the gRPC stream with standard methods.
+func NewStreamRW(stream streamSource) *StreamRW {
+	return &StreamRW{stream: stream}
+}
+
+type StreamRW struct {
+	stream streamSource
+	last   []byte
+}
+
+// Write implements io.Writer.
+func (s *StreamRW) Write(p []byte) (int, error) {
+	err := s.stream.Send(&ByteMessage{
+		Data: p,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("failed to write data to client: %w", err)
+	}
+	return len(p), nil
+}
+
+// Read implements io.Reader.
+func (s *StreamRW) Read(p []byte) (int, error) {
+	l := 0
+	if len(s.last) > 0 {
+		l = copy(p, s.last)
+	}
+
+	msg, err := s.stream.Recv()
+	if err != nil {
+		return 0, err
+	}
+
+	s.last = msg.GetData()
+	n := copy(p, s.last)
+	s.last = s.last[n:]
+
+	return n + l, nil
+}
+
+// parseHeader parses the incoming request header and extracts the method, path,
+// & header values.
+func parseHeader(r io.Reader) (method string, path string, header http.Header, err error) {
```

Comment:
so you've created your own webserver I see? interesting :)

(if it works, it works.)